src/hotspot/share/classfile/javaClasses.cpp

```diff
@@ -2667,6 +2667,41 @@ void java_lang_Throwable::get_stack_trace_elements(Handle throwable,
   }
 }
 
+objArrayOop java_lang_Throwable::get_stack_trace(Handle throwable, TRAPS) {
+  // Call to JVM to fill in the stack trace and clear declaringObject to not keep classes alive
+  // in the stack trace.
+  // call this:  public StackTraceElement[] getStackTrace()
+  assert(throwable.not_null(), "shouldn't be");
+
+  TempNewSymbol sym = SymbolTable::new_symbol("getStackTrace");
+  TempNewSymbol sig = SymbolTable::new_symbol("()[Ljava/lang/StackTraceElement;");
+
+  JavaValue result(T_ARRAY);
+  JavaCalls::call_virtual(&result, throwable,
+                          vmClasses::Throwable_klass(),
+                          sym, sig,
+                          CHECK_NULL);
+  oop stack_trace = result.get_oop();
+  assert(stack_trace->is_objArray(), "Should be an array");
+  return (objArrayOop)stack_trace;
+}
+
+oop java_lang_Throwable::recreate_cause(Symbol* exception, Symbol* message, const char* thread_name,
+                                        Handle stack_trace, TRAPS) {
+  ResourceMark rm(THREAD);
+  stringStream st;
+  if (message == NULL) {
+    st.print("in thread %s", thread_name);
+  } else {
+    st.print("%s in thread %s", message->as_C_string(), thread_name);
+  }
+  Handle h_cause = Exceptions::new_exception(THREAD, exception, st.as_string());
```

Comment:
This can fail for many reasons, causing the returned exception to not be an instance of "exception" at all, but an exception that was thrown trying to create an instance of  "exception". If that happens we should just return NULL here.

Reply:
Yes, regarding your other comment about whether the original exception should be reloaded and reinitialized, and whether its initialization could be the cause of the original exception.  I'll try to find a creative solution here so that the most common cases are still supported because I think this would be a helpful feature most of the time.  If you have some ideas, let me know (either here or offline).  Thanks.

Reply:
I wonder if I can use a predefined exception like ExceptionInInitializerError and put the original exception in with the message?

Reply:
I have to wonder whether `new_exceptions`'s behaviour is what anyone would want! It could lead to some very strange exceptions being thrown - see for example `ConstantPool::throw_resolution_error` where it can throw a resolution error with the wrong cause!

It may be okay to restrict re-creation of the exception to types in java.base as their construction seems benign with no context dependency and they should always have the necessary constructor.

Using ExceptionInInitializerError as the cause of the NCDFE may be okay for other cases.

Reply:
Thanks David.  I filed an RFE for someone to look at this (maybe me) and write some tests.

I've reworked this patch a bit and will write a test for this case.  I create the new exception when I save it, which seems like a better time to do it, so I only need OopHandle in the table. This is closer to my first prototype now.  Also, if the exception's class loader isn't NULL, I'll do a switcheroo to ExceptionInInitializerError and add the name to the exception message.  Hopefully this makes this feature useful most of the time.